backends/apple/mps/operations/ConvolutionOps.mm

```diff
@@ -9,47 +9,81 @@
 using namespace torch;
 
 PyMPSGraphTensor*
-MPSGraphModule::conv2D(MPSGraphTensor* primaryTensor, MPSGraphTensor* secondaryTensor,
-                       MPSGraphTensor* biasTensor, IntArrayRef stride,
-                       IntArrayRef padding, IntArrayRef dilation, bool transpose,
-                       IntArrayRef outputPadding, int64_t groups, bool is_depthwise) {
+MPSGraphModule::conv2D(
+  MPSGraphTensor* primaryTensor,
+  MPSGraphTensor* secondaryTensor,
+  MPSGraphTensor* biasTensor,
+  IntArrayRef stride,
+  IntArrayRef padding,
+  IntArrayRef dilation,
+  bool transpose,
+  IntArrayRef outputPadding,
+  int64_t groups,
+  bool is_depthwise) {
+  TORCH_CHECK([primaryTensor.shape count] < 5, "ConvTranspose 3D is not supported on MPS delegate");
+  TORCH_CHECK([primaryTensor dataType] == MPSDataTypeFloat32 || [primaryTensor dataType] == MPSDataTypeFloat16, "ConvTranspose 3D is not supported on MPS delegate");
+
+  // Handle 1D convolution.
+  bool isConv1D = ([secondaryTensor.shape count] == 3);
+  if (isConv1D) {
+    primaryTensor = [mpsGraph expandDimsOfTensor:primaryTensor
+                                            axis:2
+                                            name:@"unsqueezeInput"];
+    secondaryTensor = [mpsGraph expandDimsOfTensor:secondaryTensor
+                                              axis:2
+                                              name:@"unsqueezeWeight"];
+    if (stride.size() == 1) {
+      stride = IntArrayRef{1, stride[0]};
+      padding = IntArrayRef{0, padding[0]};
+      dilation = IntArrayRef{1, dilation[0]};
+      outputPadding = IntArrayRef{0, outputPadding[0]};
+    }
+  }
 
   if(is_depthwise){
-    MPSGraphDepthwiseConvolution2DOpDescriptor* desc = [MPSGraphDepthwiseConvolution2DOpDescriptor
-                                    descriptorWithStrideInX:stride[0]
-                                                  strideInY:stride[1]
-                                            dilationRateInX:dilation[0]
-                                            dilationRateInY:dilation[1]
-                                                paddingLeft:padding[1]
-                                               paddingRight:padding[1]
-                                                 paddingTop:padding[0]
-                                              paddingBottom:padding[0]
-                                               paddingStyle:MPSGraphPaddingStyleExplicit
-                                                 dataLayout:MPSGraphTensorNamedDataLayoutNCHW
-                                              weightsLayout:MPSGraphTensorNamedDataLayoutOIHW];
+    MPSGraphDepthwiseConvolution3DOpDescriptor* depthWiseConv3dDescriptor =
```

Comment:
I dont follow this change. Are you mapping 2d depthwise to 3d?

Reply:
If so why?

Reply:
We map to depthwise3D (if possible) for performance, it will be faster than the regular conv kernel. Same mapping is applied in PyTorch as well (https://github.com/pytorch/pytorch/pull/94661).